src/standardize.jl

```diff
@@ -0,0 +1,124 @@
+    μ = vec(mean(X, M))
+    σ = vec(std(X, M))
+    standardize!(X, μ, σ, obsdim)
+end
+
+function standardize!(X::AbstractVector, ::ObsDim.Constant{1})
+    μ = mean(X)
+    σ = std(X)
+    for i in 1:length(X)
+        X[i] = (X[i] - μ) / σ
+    end
+    μ, σ
+end
+
+function standardize!(X::AbstractMatrix, μ::AbstractVector, σ::AbstractVector, ::ObsDim.Constant{2})
+    σ[σ .== 0] = 1
+    nVars, nObs = size(X)
+    for iObs in 1:nObs
+        @inbounds for iVar in 1:nVars
+            X[iVar, iObs] = (X[iVar, iObs] - μ[iVar]) / σ[iVar]
+        end
+    end
+    μ, σ
+end
+
+function standardize!(X::AbstractMatrix, μ::AbstractVector, σ::AbstractVector, ::ObsDim.Constant{1})
+    σ[σ .== 0] = 1
+    nObs, nVars = size(X)
+    for iVar in 1:nVars
+        @inbounds for iObs in 1:nObs
+            X[iObs, iVar] = (X[iObs, iVar] - μ[iVar]) / σ[iVar]
+        end
+    end
+    μ, σ
+end
+
+function standardize!(X::AbstractVector, μ::AbstractVector, σ::AbstractVector, ::ObsDim.Constant{1})
+    @inbounds for i in 1:length(X)
+        X[i] = (X[i] - μ[i]) / σ[i]
+    end
+    μ, σ
+end
+
+function standardize!(X::AbstractVector, μ::AbstractFloat, σ::AbstractFloat, ::ObsDim.Constant{1})
+    @inbounds for i in 1:length(X)
+        X[i] = (X[i] - μ) / σ
+    end
+    μ, σ
+end
+
+immutable StandardScaler 
+    offset::Vector{Float64}
+    scale::Vector{Float64}
+    obsdim::ObsDim.Constant{}
+
+    function StandardScaler(offset, scale, obsdim)
+        @assert length(offset) == length(scale) 
+        new(offset, scale, convert(ObsDimension, obsdim))
+    end
+end
+
+function StandardScaler{T<:Real}(X::AbstractMatrix{T}; obsdim=LearnBase.default_obsdim(X))
+    StandardScaler(X, convert(ObsDimension, obsdim))
+end
+
+function StandardScaler{T<:Real,M}(X::AbstractArray{T,M}, ::ObsDim.Last)
+    StandardScaler(X, ObsDim.Constant{M})
+end
+
+function StandardScaler{T<:Real,M}(X::AbstractArray{T,M}, obsdim::ObsDim.Constant{M})
+    StandardScaler(mean(X, M), std(X, M), obsdim)
+end
+
+function StandardScaler{T<:Real,M}(X::AbstractArray{T,M}, offset, scale; obsdim=LearnBase.default_obsdim(X))
+    StandardScaler(offset, scale, convert(ObsDimension, obsdim))
+end
+
+function StandardScaler{T<:Real,M}(X::AbstractArray{T,M}, offset, scale, ::ObsDim.Last)
+    StandardScaler(offset, scale, ObsDim.Constant{M})
+end
+
+function StatsBase.fit{T<:Real}(::Type{StandardScaler}, X::AbstractMatrix{T}; obsdim=LearnBase.default_obsdim(X))
+    StandardScaler(X, obsdim=obsdim)
+end
+
+function transform!{T<:Real}(cs::StandardScaler, X::AbstractMatrix{T})
+    @assert length(cs.offset) == size(X, 1)
+    standardize!(X, cs.offset, cs.scale, obsdim=cs.obsdim)
+    X
+end
+
+function transform{T<:AbstractFloat}(cs::StandardScaler, X::AbstractMatrix{T})
+    Xnew = copy(X)
+    transform!(cs, Xnew)
+end
+
+function transform{T<:Real}(cs::StandardScaler, X::AbstractMatrix{T})
+    X = convert(AbstractMatrix{AbstractFloat}, X)
```

Comment:
Avoid using abstract types as storage. Especially when working with arrays. 

This resulting array will be painful to work with and really slow

```julia
julia> convert(AbstractMatrix{AbstractFloat}, rand(2,2))
2×2 Array{AbstractFloat,2}:
 0.468758  0.413961
 0.559098  0.174936
```

This video explains pretty well the difference between abstract and concrete element types (I tagged the beginning of the relevant section). The audio is a bit out of sync with the video though.

https://youtu.be/ag_NtJRmYg8?t=12m36s